teps/0126-allow-task-sidecars-to-be-specified-in-pipelinerun.md

```diff
@@ -0,0 +1,123 @@
+
+Workaround #2 is not seen as adequate because the parallel task has a lifetime longer than the task run, a finally task is needed to terminate it, and it uses the feature "parallel tasks" to mimick another feature ("sidecars") instead of giving access to the sidecar feature where the user needs it.
+
+Creating a task with a sidecar baked in (workaround #3) certainly works, but it prevents reuse: I can't use the `golang-test` task anymore just because there is no sidecar defined. The task I would need to create is exactly the same as `golang-test`, except for the sidecar definition. If there was a way to override sidecars, I could use the task from the ArtifactHub as is and supply the sidecar configuration fitting for the pipeline run.
+
+## Proposal
+
+Users would be able to define the following PipelineRun:
+
+```
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: example 
+spec:
+  tasks:
+  - name: test
+    taskRef:
+      resolver: hub
+      params:
+      - name: kind
+        value: task
+      - name: name
+        value: golang-test
+      - name: version
+        value: "0.6"
+    sidecars:
+    - image: postgres:14
+      name: postgres
+      env:
+      - name: POSTGRES_DB
+        value: foobar
+    workspaces:
+    - name: source
+      persistentVolumeClaim:
+        claimName: my-source
+```
+
+This would behave exactly the same as if the `golang-test` task had specified the sidecar itself.
+
+### Notes and Caveats
+
+I see a bit of conflict between this proposal and the design of [Overriding Task Steps and Sidecars](https://tekton.dev/docs/pipelines/taskruns/#overriding-task-steps-and-sidecars). One might think that `sidecarOverrides` would allow to override which sidecars are used and what their image is etc., however the feature is restricted to `resources` only, at least for now.
+
+If `sidecarOverrides` would allow to override more fields (in particular, `image` and `env` seem crucial) this may solve part of the use case explained above. However, I do not see how the design of `sidecarOverrides` could be used to e.g. specify sidecars when the task itself does not define any, or add a second sidecar if the task defines one sidecar. Therefore I would see the proposal here separate from `sidecarOverrides` even though some of the motivation is shared.
+
+In any case, the proposed solution must accomodate the situation where the task already defines one or more sidecars. Then, it must be clear to pipeline run authors how the feature of "specifying / overriding sidecars in pipeline runs" interacts with task-defined sidecars. I see the option to: (a) override the task definition, (b) extend the task defition or (c) allow both, e.g. via different fields. It is not clear for me (yet) what the best option would be.
```

Comment:
Also, what would happens in `sidecars` from a `PipelineRun` comes with the exact same name as an already defined sidecar in the `Task` definition.

Reply:
I guess in such case we should probably fail?